packages/backend-common/src/appContext/ApplicationContext.ts

```diff
@@ -0,0 +1,185 @@
+    identifier,
+  }: {
+    logger: Logger;
+    dependencies: AnyDependencyConfig[];
+    container?: interfaces.Container;
+    identifier?: string;
+  }) {
+    logger.info(
+      `Constructing Inversify root container with dependencies ${[
+        ...new Set(dependencies.map(it => it.id)),
+      ].join(', ')}`,
+    );
+    const boundContainer =
+      InversifyApplicationContext.bindDependenciesIfNotBound(
+        dependencies,
+        container ?? new Container(),
+      );
+    return new InversifyApplicationContext({
+      identifier: identifier ?? 'root',
+      container: boundContainer,
+      logger,
+    });
+  }
+
+  private constructor({
+    identifier,
+    container,
+    childContexts,
+    logger,
+  }: {
+    identifier: string;
+    container: interfaces.Container;
+    childContexts?: Map<string, ApplicationContext>;
+    logger: Logger;
+  }) {
+    this.identifier = identifier;
+    this.container = container;
+    this.logger = logger;
+    this.childContexts = childContexts ?? new Map<string, ApplicationContext>();
+  }
+
+  getContainer() {
+    return this.container;
+  }
+
+  get<T>(dep: Dependency<T>) {
+    return this.container.get<T>(dep.id);
+  }
+
+  getChildContext(identifier: string): ApplicationContext {
+    if (this.childContexts.has(identifier)) {
+      return this.childContexts.get(identifier)!;
+    }
+    throw Error(
+      `Unable to find application context for identifier ${identifier}`,
+    );
+  }
+
+  createBoundPlugin({
+    plugin,
+    createRouter,
+    dependencies,
+  }: BoundPluginOptions): BoundPlugin {
+    this.logger.info(
+      `Creating application context for plugin '${
+        this.identifier
+      }'. Binding dependencies ${[
+        ...new Set(dependencies.map(it => it.id)),
+      ].join(', ')}.`,
+    );
+    const childContainer = this.getContainer();
+    InversifyApplicationContext.bindDependenciesIfNotBound(
+      dependencies,
+      childContainer,
+    );
+    return {
+      name: plugin,
+      router: createRouter(this),
+      getDependencies() {
+        return dependencies.map(it => it.id);
+      },
+    };
+  }
+
+  forPlugin(plugin: string): ApplicationContext {
+    const childContext = new InversifyApplicationContext({
+      identifier: plugin,
+      container: this.container.createChild(),
+      logger: this.logger.child({ type: 'plugin', plugin }),
+    });
+    this.childContexts.set(plugin, childContext);
+    return childContext;
+  }
+
+  private static bindDependenciesIfNotBound(
+    configs: AnyDependencyConfig[],
+    container: interfaces.Container,
+  ) {
+    configs.forEach(dependencyConfig => {
+      if (!container.isBound(dependencyConfig.id.id)) {
+        container
+          .bind(dependencyConfig.id.id)
+          // TODO: We are always binding to a dynamic value here by calling the factory to construct a dependency.
+          // We should accept a config option in dependencyConfig to determine the binding type instead
```

Comment:
Missing constants, async providers etc. that Inversify supports.

Reply:
This should also check the correct binding level for cases where we want to force a binding for the child container

Reply:
Checking binding level is not really relevant anymore here, since we are shipping an individual container for each module. Adding consts/providers etc. can be done on later stages when there is need for those. Currently we don't have any of those yet.